native/src/main/java/io/ballerina/stdlib/serdes/BallerinaStructuredTypeMessageSerializer.java

```diff
@@ -82,55 +83,62 @@ public Builder generateMessage() {
-                    break;
-                case TypeTags.MAP_TAG:
-                    @SuppressWarnings("unchecked")
-                    BMap<BString, Object> mapValue = (BMap<BString, Object>) ballerinaValue;
-                    messageSerializer.setMapFieldValue(mapValue);
-                    break;
-                default:
-                    throw createSerdesError(UNSUPPORTED_DATA_TYPE + ballerinaType.getName(), SERDES_ERROR);
-            }
+            setFieldValue(ballerinaValue, ballerinaType);
         }
 
         return messageSerializer.getDynamicMessageBuilder();
     }
+
+    private void setFieldValue(Object ballerinaValue, Type ballerinaType) {
+        switch (ballerinaType.getTag()) {
+            case TypeTags.NULL_TAG:
+                messageSerializer.setNullFieldValue(ballerinaValue);
+                break;
+            case TypeTags.INT_TAG:
+                messageSerializer.setIntFieldValue(ballerinaValue);
+                break;
+            case TypeTags.BYTE_TAG:
+                messageSerializer.setByteFieldValue((Integer) ballerinaValue);
+                break;
+            case TypeTags.FLOAT_TAG:
+                messageSerializer.setFloatFieldValue(ballerinaValue);
+                break;
+            case TypeTags.STRING_TAG:
+                messageSerializer.setStringFieldValue((BString) ballerinaValue);
+                break;
+            case TypeTags.BOOLEAN_TAG:
+                messageSerializer.setBooleanFieldValue((Boolean) ballerinaValue);
+                break;
+            case TypeTags.DECIMAL_TAG:
+                messageSerializer.setDecimalFieldValue((BDecimal) ballerinaValue);
+                break;
+            case TypeTags.UNION_TAG:
+                messageSerializer.setUnionFieldValue(ballerinaValue);
+                break;
+            case TypeTags.ARRAY_TAG:
+                messageSerializer.setArrayFieldValue((BArray) ballerinaValue);
+                break;
+            case TypeTags.RECORD_TYPE_TAG:
+                @SuppressWarnings("unchecked")
+                BMap<BString, Object> recordValue = (BMap<BString, Object>) ballerinaValue;
+                messageSerializer.setRecordFieldValue(recordValue);
+                break;
+            case TypeTags.TUPLE_TAG:
+                messageSerializer.setTupleFieldValue((BArray) ballerinaValue);
+                break;
+            case TypeTags.TABLE_TAG:
+                messageSerializer.setTableFieldValue((BTable<?, ?>) ballerinaValue);
+                break;
+            case TypeTags.MAP_TAG:
+                @SuppressWarnings("unchecked")
+                BMap<BString, Object> mapValue = (BMap<BString, Object>) ballerinaValue;
+                messageSerializer.setMapFieldValue(mapValue);
+                break;
+            case TypeTags.TYPE_REFERENCED_TYPE_TAG:
+                setFieldValue(ballerinaValue, ((ReferenceType) ballerinaType).getReferredType());
+                break;
```

Comment:
Is this the correct way to achieve this? Shouldnt the type for refernece type resolved before the setFieldType method itself?

Reply:
As we are using a `switch-case` to call the suitable `setFieldValue()` method for the type, I think resolving it earlier will also have the same behavior.
And handling it in the switch case might give a small performance advantage for recursive calls.
 
Btw, is it possible to get an `IntersectionType` as a field type here? @MohamedSabthar

Reply:
```
if type is referencetype
   // recursive get the base type
else
    continue;
   
Switch for set field
```

Checking once if it is a refernce type should be enough. There would be performance hit if this is a refernce type, it is 11 switch cases for recursive reference type. Or else if this is a refernce type, we recursively resolve it and call the setFieldMethod

Reply:
I'll update the PR